### PR TITLE
build: treat incompatible pointer warning as error

### PR DESCRIPTION
**Pull request by ObserverOfTime**
See #3779

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -135,26 +135,41 @@ jobs:
...
 
     - name: Build C library (CMake)
       if: ${{ !matrix.use-cross }}
-      run: |-
+      run: |
         cmake -S lib -B build \
           -DBUILD_SHARED_LIBS=OFF \
           -DCMAKE_BUILD_TYPE=Debug \
-          -DCMAKE_COMPILE_WARNING_AS_ERROR=ON
+          -DCMAKE_C_COMPILER=clang \
```
Comment (ObserverOfTime) [inline .github/workflows/build.yml:159]: Using clang to avoid an [incorrect GCC diagnostic](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=83584).

**Comment by ObserverOfTime** [comment]
Is this supposed to be used anywhere?
https://github.com/tree-sitter/tree-sitter/blob/178c5d93f4ca8c962dc977365148148edbce1f24/lib/src/wasm_store.c#L162

**Comment by ObserverOfTime** [comment]
I don't understand what's wrong with the linker…

**Comment by amaanq** [comment]
> Is this supposed to be used anywhere?
> 
> https://github.com/tree-sitter/tree-sitter/blob/178c5d93f4ca8c962dc977365148148edbce1f24/lib/src/wasm_store.c#L162

:thinking: don't think so, and looks like it was added in 2022, so my guess is it's a remnant of the code from back then and can be removed.

File: lib/src/tree_cursor.c [hunk 2]
```diff
@@ -253,7 +256,11 @@ bool ts_tree_cursor_goto_last_child(TSTreeCursor *self) {
         return false;
     }
   }
-  return false;
+#ifdef _MSC_VER
+  __assume(0);
+#else
+  __builtin_unreachable();
+#endif
```
Comment (amaanq) [inline lib/src/tree_cursor.c:263]: Can we revert this and ignore the warning MSVC unnecessarily complains about?

Comment (ObserverOfTime) [reply lib/src/tree_cursor.c:263]: I mean this is technically an optimisation.

Comment (amaanq) [reply lib/src/tree_cursor.c:263]: a compiler from this millennia would know that this is unreachable code regardless

**Comment by BobbieCarroll** [comment]
Redownload for a interrupt for the first time you have the right access to
my account and the app says it's not working with me and I'm just not
interested in the fact of other stuff.

**Pull request merged**